twenty-first/src/math/polynomial.rs

```diff
@@ -1360,6 +1416,403 @@ where
+            .map(|(lc, sq)| sq.scalar_mul(FF::from(lc.value())) - Polynomial::one())
+            .collect_vec();
+
+        // precompute NTT-friendly multiple of the modulus
+        let (shift_coefficients, tail_length) = modulus.shift_factor_ntt_with_tail_size();
+
+        ModularInterpolationPreprocessingData {
+            even_zerofiers,
+            odd_zerofiers,
+            shift_coefficients,
+            tail_length,
+        }
+    }
+
+    /// Compute f(X) mod m(X) where m(X) is a given modulus and f(X) is the
+    /// interpolant of a list of n values on a domain which is a coset of the
+    /// size-n subgroup that is identified by some offset.
+    fn fast_modular_coset_interpolate(
+        values: &[FF],
+        offset: BFieldElement,
+        modulus: &Polynomial<FF>,
+    ) -> Self {
+        let preprocessing_data =
+            Self::fast_modular_coset_interpolate_preprocess(values.len(), offset, modulus);
+        Self::fast_modular_coset_interpolate_with_zerofiers_and_ntt_friendly_multiple(
+            values,
+            offset,
+            modulus,
+            &preprocessing_data,
+        )
+    }
+
+    /// Only marked `pub` for benchmarking purposes. Not considered part of the
+    /// interface.
+    #[doc(hidden)]
+    pub fn fast_modular_coset_interpolate_with_zerofiers_and_ntt_friendly_multiple(
+        values: &[FF],
+        offset: BFieldElement,
+        modulus: &Polynomial<FF>,
+        preprocessed: &ModularInterpolationPreprocessingData<FF>,
+    ) -> Self {
+        if modulus.is_zero() {
+            panic!("cannot reduce modulo zero")
+        };
+        let n = values.len();
+        let omega = BFieldElement::primitive_root_of_unity(n as u64).unwrap();
+
+        if n < Self::FAST_MODULAR_COSET_INTERPOLATE_CUTOFF_THRESHOLD_PREFER_LAGRANGE {
+            let domain = (0..n)
+                .scan(FF::from(offset.value()), |acc: &mut FF, _| {
+                    let yld = *acc;
+                    *acc *= omega;
+                    Some(yld)
+                })
+                .collect::<Vec<FF>>();
+            let interpolant = Self::lagrange_interpolate(&domain, values);
+            return interpolant.reduce(modulus);
+        } else if n <= Self::FAST_MODULAR_COSET_INTERPOLATE_CUTOFF_THRESHOLD_PREFER_INTT {
+            let mut coefficients = values.to_vec();
+            intt(&mut coefficients, omega, n.ilog2());
+            let interpolant = Polynomial::new(coefficients);
+
+            return interpolant
+                .scale(FF::from(offset.inverse().value()))
+                .reduce_by_ntt_friendly_modulus(
+                    &preprocessed.shift_coefficients,
+                    preprocessed.tail_length,
+                )
+                .reduce(modulus);
+        }
+
+        // Use even-odd domain split.
+        // Even: {offset * omega^{2*i} | i in 0..n/2}
+        // Odd: {offset * omega^{2*i+1} | i in 0..n/2} = {(offset * omega) * omega^{2*i} | i in 0..n/2}
```

Comment:
Violates the 100 character line width and the 80 character comment-length width.

```suggestion
        // Odd: {offset * omega^{2*i+1} | i in 0..n/2}
        //      = {(offset * omega) * omega^{2*i} | i in 0..n/2}
```

Reply:
42999c2f